tony-core/src/main/java/com/linkedin/tony/TonyApplicationMaster.java

```diff
@@ -93,7 +99,9 @@
   private String amHostPort;
   private static FileSystem fs;
   private static String tonyHistoryFolder;
+  private static int writeRetryCnt;
   private static Path jobDir;
+  private Queue<Event> eventQueue = new LinkedList<>();
```

Comment:
As Anthony mentioned earlier, this should be a BlockingQueue, and a separate thread can write the events to the history file while the application is running. Since the events are written concurrently, it won't take so much time at the end to write the history file if there are a lot of events. Also, in case of failure, there would be some information in the log, which will be useful for debugging the failure.

The history writing logic can be put in a separate class (HistoryWriter, EventLogger), and the queue could be moved there. It would be started and stopped by the ApplicationMaster, and have methods for enqueuing events.